### ISSUE TITLE
Stack overflow when deserializing recursive Rc

### ISSUE THREAD

**Issue by 5225225** [rkyv/rkyv#301]
```rust
use std::rc::Rc;

#[derive(rkyv::Archive, rkyv::Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[archive_attr(derive(bytecheck::CheckBytes, PartialEq, Eq, PartialOrd, Ord))]
#[archive_attr(check_bytes(bound = "__C: rkyv::validation::ArchiveContext + rkyv::validation::SharedContext, <__C as rkyv::Fallible>::Error: std::error::Error"))]
#[archive(bound(deserialize = "__D: rkyv::de::SharedDeserializeRegistry"))]
enum AllValues {
    Rc( #[omit_bounds] #[archive_attr(omit_bounds)] Rc<AllValues>),
}

fn main() {
    let data = [251, 0, 255, 255, 0, 0, 0, 255, 252, 255, 255, 255, 1, 0, 0, 0, 0, 0, 246, 255,
    240, 255, 255, 255];
    rkyv::from_bytes::<AllValues>(&data);
}
```

Address Sanitizer says:

```rust
=================================================================
==534093==ERROR: AddressSanitizer: stack-overflow on address 0x7ffcd209fff8 (pc 0x5588b8067258 bp 0x7ffcd20a00f0 sp 0x7ffcd20a0000 T0)
    #0 0x5588b8067258 in _$LT$rkyv..de..deserializers..alloc..SharedDeserializeMap$u20$as$u20$rkyv..de..SharedDeserializeRegistry$GT$::get_shared_ptr::hfbbc5b6093048f1c (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/fuzz_target_1+0x21c258)
    #1 0x5588b7fe269d in rkyv::de::SharedDeserializeRegistry::deserialize_shared::hea3c7a1b4d16afd4 (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/fuzz_target_1+0x19769d)
    #2 0x5588b7fe2743 in rkyv::de::SharedDeserializeRegistry::deserialize_shared::hea3c7a1b4d16afd4 (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/fuzz_target_1+0x197743)
    #3 0x5588b7fe2743 in rkyv::de::SharedDeserializeRegistry::deserialize_shared::hea3c7a1b4d16afd4 (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/fuzz_target_1+0x197743)
    #4 0x5588b7fe2743 in rkyv::de::SharedDeserializeRegistry::deserialize_shared::hea3c7a1b4d16afd4 (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/fuzz_target_1+0x197743)
[...]
    #245 0x5588b7fe2743 in rkyv::de::SharedDeserializeRegistry::deserialize_shared::hea3c7a1b4d16afd4 (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/fuzz_target_1+0x197743)
    #246 0x5588b7fe2743 in rkyv::de::SharedDeserializeRegistry::deserialize_shared::hea3c7a1b4d16afd4 (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/fuzz_target_1+0x197743)

SUMMARY: AddressSanitizer: stack-overflow (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/fuzz_target_1+0x21c258) in _$LT$rkyv..de..deserializers..alloc..SharedDeserializeMap$u20$as$u20$rkyv..de..SharedDeserializeRegistry$GT$::get_shared_ptr::hfbbc5b6093048f1c
==534093==ABORTING
```

**Comment by djkoloski** [comment]
The default validator has a maximum subtree depth of `usize::MAX`, but it seems like a good idea to be cap it by default. You can test with a maximum subtree depth using:

```rust
use rkyv::de::deserializers::SharedDeserializeMap;
use rkyv::validation::{check_archived_root_with_context, validators::ArchiveValidator};

const MAX_DEPTH: usize = 50;
let value = check_archived_root_with_context(
    &data,
    &mut ArchiveValidator::with_max_depth(&data, MAX_DEPTH),
);
// then deserialize
let deserialized = value.deserialize(&mut SharedDeserializeMap::new());
```

There's probably also be a case here where `Rc` validation shouldn't be recursing. #214 also relates to object cycles.

**Comment by djkoloski** [comment]
The exact issue here was fixed in 0a3a6906f8e99db00a254da12628890f3b59cd29 by detecting reference cycles during deserialization. However, cyclic shared pointers could still pass validation, which is a sharp edge for anyone who wants to accept potentially-malicious data. To mitigate that problem, 59c5c4032645dbf40237d2809b91345f5157413d introduces some extra machinery to allow users to detect and error on cyclic shared pointers. If anyone wants to access data that does contain cyclic shared pointers, they can make a few modifications.
